pkg/tbtc/dkg.go

```diff
@@ -480,47 +484,180 @@ func (de *dkgExecutor) registerSigner(
+			// need to add 1.
+			memberIndexes = append(memberIndexes, group.MemberIndex(index+1))
+		}
+	}
+
+	if len(memberIndexes) == 0 {
+		dkgLogger.Infof(
+			"not eligible for DKG result approval; my operator "+
+				"ID [%v] is not among DKG participants [%v]",
+			operatorID,
+			result.Members,
+		)
+		return
+	}
+
+	dkgLogger.Infof("scheduling DKG result approval")
+
+	parameters, err := de.chain.DKGParameters()
+	if err != nil {
+		dkgLogger.Errorf("cannot get current DKG parameters: [%v]", err)
+		return
+	}
+
+	// The challenge period starts at the result submission block and lasts
+	// for challengePeriodBlocks.
+	challengePeriodEndBlock := submissionBlock + parameters.ChallengePeriodBlocks
+	// The approval is possible one block after the challenge period end.
+	// The result submitter has precedence for approvePrecedencePeriodBlocks.
+	approvePrecedencePeriodStartBlock := challengePeriodEndBlock + 1
+	// Everyone else can approve once the precedence period ends.
+	approvePeriodStartBlock := approvePrecedencePeriodStartBlock +
+		parameters.ApprovePrecedencePeriodBlocks
+
+	for _, currentMemberIndex := range memberIndexes {
+		go func(memberIndex group.MemberIndex) {
+			var approveBlock uint64
+
+			if memberIndex == result.SubmitterMemberIndex {
+				// The submitter can approve earlier, during the precedence
+				// period.
+				approveBlock = approvePrecedencePeriodStartBlock
+			} else {
+				// Everyone else must approve after the precedence period ends.
+				// Each member preserves a delay according to their index
+				// to avoid simultaneous approval.
+				delayBlocks := uint64(memberIndex-1) * dkgResultApprovalDelayStepBlocks
+				approveBlock = approvePeriodStartBlock + delayBlocks
+			}
+
+			dkgLogger.Infof(
+				"[member:%v] waiting for block [%v] to approve DKG result",
+				memberIndex,
+				approveBlock,
+			)
+
+			ctx, cancelCtx := context.WithCancel(context.Background())
+			defer cancelCtx()
+
+			subscription := de.chain.OnDKGResultApproved(
+				func(event *DKGResultApprovedEvent) {
+					cancelCtx()
+				},
+			)
+			defer subscription.Unsubscribe()
+
+			err := de.waitForBlockFn(ctx, approveBlock)
+			if err != nil {
+				dkgLogger.Errorf(
+					"[member:%v] error while waiting for DKG result "+
+						"approve block: [%v]",
+					memberIndex,
+					err,
+				)
+				return
+			}
+
+			// If the context got cancelled that means the result was approved
+			// by someone else.
+			if ctx.Err() != nil {
+				dkgLogger.Infof(
+					"[member:%v] DKG result approved by someone else",
+					memberIndex,
+				)
+				return
+			}
+
+			err = de.chain.ApproveDKGResult(result)
+			if err != nil {
+				dkgLogger.Errorf(
+					"[member:%v] cannot approve DKG result: [%v]",
+					memberIndex,
+					err,
+				)
+				return
+			}
+
+			dkgLogger.Infof("[member:%v] approving DKG result", memberIndex)
+		}(currentMemberIndex)
+	}
 }
 
 // finalSigningGroup takes three parameters:
```

Comment:
Triple-checking this part. It this all correct?

1. Signing group members are selected on-chain. Some of them may be inactive or malicious and excluded from the given attempt during the announcement phase or by the retry loop.
2. During the execution of the off-chain key generation protocol they always have the same member indexes as their positions in the signing group selected on-chain.
3. The member index is used in the cryptography key generation protocol.
4. Once we are done with the protocol, we shift some member indexes before saving them on disk so that no empty places are left after inactive/misbehaving members that would lead to delayed submission of work to the chain.
5. We can do it because the member index no longer matters for the signing protocol. The original member index was saved in TSS data specific to the given key share.

I also wonder if it makes sense to improve `finalSigningGroup` documentation by including all these points.

Reply:
> Once we are done with the protocol, we shift some member indexes before saving them on disk **so that no empty places are left after inactive/misbehaving members that would lead to delayed submission of work to the chain.**

The bold part is not true. Indeed, we shift but just to have proper indexes for signing. For result submission, we still need to operate on original indexes as the contract does some validations of group membership based on them.

Reply:
By *delayed submission of work to the chain*  I did not mean the DKG result submission because for this one, as you say, we need to use the original member index from the group selection, but everything that will potentially happen later: inactivity claim or one or two Bridge functions only group member can call.

Does it make sense now?

Also, regarding signing, how do we use the member index for signing? My understanding is that we could not shift the indexes as the index used for cryptography protocol in tss-lib is stored in `*tecdsa.PrivateKeyShare` and this is the original one from the group selection. The member index shifting is just to not have gaps in logs when it comes to signing. For example if we have a group of 95 members, we want the indexes to be from 1-95.

Is this correct?

Reply:
> By delayed submission of work to the chain I did not mean the DKG result submission because for this one, as you say, we need to use the original member index from the group selection, but everything that will potentially happen later: inactivity claim or one or two Bridge functions only group member can call.
>
> Does it make sense now?

Yep, perfectly makes sense.

> Also, regarding signing, how do we use the member index for signing? My understanding is that we could not shift the indexes as the index used for cryptography protocol in tss-lib is stored in *tecdsa.PrivateKeyShare and this is the original one from the group selection. The member index shifting is just to not have gaps in logs when it comes to signing. For example if we have a group of 95 members, we want the indexes to be from 1-95.
>
> Is this correct?

If we end up with 95 members as a result of DKG, the underlying indexes in the key shares are in the same range, so 1-95. Shifting, however, is not only about log gaps but mostly about how the signing algorithm selects the members for the given attempt. It needs to pick up 51 members out of X (actual group size). If we don't shift, X is 95, and we have excluded 5 members with indexes <= 95, the algorithm may select those excluded members and will never select proper members with indexes > 95. This is the main reason

Reply:
> If we end up with 95 members as a result of DKG, the underlying indexes in the key shares are in the same range, so 1-95. 

Can you please point me to the place in the code where this is happening? I am struggling to find it.

Reply:
Wait, I was wrong saying this: "If we end up with 95 members as a result of DKG, the underlying indexes in the key shares are in the same range, so 1-95". Those are original indexes from group selection as you said, see [here](https://github.com/keep-network/keep-core/blob/b8dcba4a37e68febff082bd2c6282b5fac1f7d04/pkg/tecdsa/dkg/member.go#L132). However, the rest I said above is still true.

Reply:
All conversions between our member index and the underlying index from the key share (TSS id) are then handled using `identityConverter` that has two implementations, one for DKG and one for signing.

Reply:
I see we have all of it explained perfectly in the identify converter docs:

https://github.com/keep-network/keep-core/blob/b8dcba4a37e68febff082bd2c6282b5fac1f7d04/pkg/tecdsa/common/common.go#L14-L60

Can we please just drop one sentence to the end of `finalSigningGroup` docs to avoid having this discussion next time? Something like:
```
// Please see docs of identityConverter from pkg/tecdsa/common/ for more
// information about shifting indexes.
```

Reply:
See de2d1826